flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/TypeInference.java

```diff
@@ -0,0 +1,183 @@
+
+	private TypeInference(
+			InputTypeValidator inputTypeValidator,
+			TypeStrategy accumulatorTypeStrategy,
+			TypeStrategy outputTypeStrategy,
+			@Nullable List<String> argumentNames,
+			@Nullable List<DataType> argumentTypes) {
+		this.inputTypeValidator = inputTypeValidator;
+		this.accumulatorTypeStrategy = outputTypeStrategy;
+		this.outputTypeStrategy = outputTypeStrategy;
+		if (argumentNames != null && argumentTypes != null && argumentNames.size() != argumentTypes.size()) {
+			throw new IllegalArgumentException(
+				String.format(
+					"Mismatch between argument types %d and argument names %d.",
+					argumentNames.size(),
+					argumentTypes.size()));
+		}
+		this.argumentNames = argumentNames;
+		this.argumentTypes = argumentTypes;
+	}
+
+	public InputTypeValidator getInputTypeValidator() {
+		return inputTypeValidator;
+	}
+
+	public TypeStrategy getAccumulatorTypeStrategy() {
+		return accumulatorTypeStrategy;
+	}
+
+	public TypeStrategy getOutputTypeStrategy() {
+		return outputTypeStrategy;
+	}
+
+	public Optional<List<String>> getArgumentNames() {
+		return Optional.ofNullable(argumentNames);
+	}
+
+	public Optional<List<DataType>> getArgumentTypes() {
+		return Optional.ofNullable(argumentTypes);
+	}
+
+	// --------------------------------------------------------------------------------------------
+
+	/**
+	 * Builder for configuring and creating instances of {@link TypeInference}.
+	 */
+	public static class Builder {
+
+		private InputTypeValidator inputTypeValidator = InputTypeValidators.PASSING;
+
+		private @Nullable TypeStrategy accumulatorTypeStrategy;
+
+		private @Nullable TypeStrategy outputTypeStrategy;
+
+		private @Nullable List<String> argumentNames;
+
+		private @Nullable List<DataType> argumentTypes;
+
+		public Builder() {
+			// default constructor to allow a fluent definition
+		}
+
+		/**
+		 * Sets the validator for checking the input data types of a function call.
+		 *
+		 * <p>A always passing function is assumed by default (see {@link InputTypeValidators#PASSING}).
+		 */
+		public Builder inputTypeValidator(InputTypeValidator inputTypeValidator) {
+			this.inputTypeValidator =
+				Preconditions.checkNotNull(inputTypeValidator, "Input type validator must not be null.");
+			return this;
+		}
+
+		/**
+		 * Sets the strategy for inferring the intermediate accumulator data type of a function call.
+		 *
+		 * <p>By default the accumulator type is assumed to be equal to the output type strategy (see
+		 * {@link #outputTypeStrategy(TypeStrategy)}).
+		 */
+		public Builder accumulatorTypeStrategy(TypeStrategy accumulatorTypeStrategy) {
+			this.accumulatorTypeStrategy =
+				Preconditions.checkNotNull(accumulatorTypeStrategy, "Accumulator type strategy must not be null.");
+			return this;
+		}
+
+		/**
+		 * Sets the strategy for inferring the final output data type of a function call.
+		 *
+		 * <p>Required.
+		 */
+		public Builder outputTypeStrategy(TypeStrategy outputTypeStrategy) {
```

Comment:
Personal taste, so feel free to ignore as we never had a proper discussion on design patterns standards.

Personally, in most cases, I prefer to have a static factory method with the required parameter(s). In this case I would prefer `public static Builder TypeInference#outputTypeStrategy(TypeStrategy outputTypeStrategy)`, but as I said this is just a personal taste.